cassandra/policies.py

```diff
@@ -864,6 +874,228 @@ def _add_jitter(self, value):
         return min(max(self.base_delay, delay), self.max_delay)
 
 
+class ShardReconnectionScheduler(ABC):
+    @abstractmethod
+    def schedule(
+            self,
+            host_id: str,
+            shard_id: int,
+            method: Callable[..., None],
+            *args: List[Any],
+            **kwargs: dict[Any, Any]) -> None:
+        raise NotImplementedError()
+
+
+class ShardReconnectionPolicy(ABC):
+    """
+    Base class for shard reconnection policies.
+
+    On `new_scheduler` instantiate a scheduler that behaves according to the policy
+    """
+
+    @abstractmethod
+    def new_scheduler(self, session: Session) -> ShardReconnectionScheduler:
+        raise NotImplementedError()
+
+
+class NoDelayShardReconnectionPolicy(ShardReconnectionPolicy):
+    """
+    A shard reconnection policy with no delay between attempts and no concurrency restrictions.
+    Ensures at most one pending reconnection per (host, shard) pair — any additional
+     reconnection attempts for the same (host, shard) are silently ignored.
+
+    On `new_scheduler` instantiate a scheduler that behaves according to the policy
+    """
+
+    def new_scheduler(self, session: Session) -> ShardReconnectionScheduler:
+        return _NoDelayShardReconnectionScheduler(session)
+
+
+class _NoDelayShardReconnectionScheduler(ShardReconnectionScheduler):
+    session: Session
+    already_scheduled: dict[str, bool]
+
+    def __init__(self, session: Session):
+        self.session = weakref.proxy(session)
+        self.already_scheduled = {}
+
+    def _execute(
+            self,
+            scheduled_key: str,
+            method: Callable[..., None],
+            *args: List[Any],
+            **kwargs: dict[Any, Any]) -> None:
+        try:
+            method(*args, **kwargs)
+        finally:
+            self.already_scheduled[scheduled_key] = False
+
+    def schedule(
+            self,
+            host_id: str,
+            shard_id: int,
+            method: Callable[..., None],
+            *args: List[Any],
+            **kwargs: dict[Any, Any]) -> None:
+        scheduled_key = f'{host_id}-{shard_id}'
+        if self.already_scheduled.get(scheduled_key):
+            return
+
+        self.already_scheduled[scheduled_key] = True
+        if not self.session.is_shutdown:
```

Comment:
For example here, in `_NoDelayShardReconnectionScheduler`: It performs the check in obviously non-thread-safe way. So if it can be called concurrently, then multiple schedules for the same key are possible, despite `already_scheduled` trying to prevent that. So now I'm thinking that maybe it can't be called concurrently?

OTOH `already_scheduled` uses a lock, which is extremely strong signal that concurrency is at play here. And now I have no idea what to think, because nothing is explained anywhere.